src/main/scala/scalaz/stream/io.scala

```diff
@@ -212,4 +217,173 @@ object io {
         else buf.take(m)
       }}
     }
+
+  /**
+   * Converts a source to a mutable `InputStream`.  The resulting input stream
+   * should be reasonably efficient and supports early termination (i.e. all
+   * finalizers associated with the input process will be run if the stream is
+   * closed).
+   *
+   * Takes a `copy` function which itself takes the following parameters in order:
+   *
+   * - `A` - the source chunk
+   * - `Int` - the logical offset into the source chunk
+   * - `Array[Byte]` - the target copy buffer
+   * - `Int` - the logical offset into the target buffer
+   * - `Int` - the length to copy
+   *
+   * The `copy` function returns the number of bytes copied, in total.  You should
+   * think of this function as a generic version of `System.arraycopy`, which is
+   * in fact the precise implementation you would use if `A = Array[Byte]`.  Note
+   * that `toInputStream` will never request a `copy` with a length value that
+   * exceeds the logical length of the source chunk as reported by the `logicalLength`
+   * function.  Additionally, `buffer` will always be *at least* `off2 + length` in
+   * size.  Thus, the `copy` function does not need to concern itself with bounds
+   * checking.
+   *
+   * The other function taken as a parameter is `logicalLength`, which simply returns
+   * the logical length of the source chunk in question.  Thus, if `A = Array[Byte]`,
+   * this function would simply be `{ _.length }`.
+   *
+   * @see #toInputStreamFromBytes
+   */
+  def toInputStream[A](p: Process[Task, A])(logicalLength: A => Int, copy: (A, Int, Array[Byte], Int, Int) => Int): InputStream = new InputStream {
```

Comment:
I think simplifying the signature to
```scala
def toInputStream(p: Process[Task, Array[Byte]]): InputStream = ???
```
would be the most usable API. Any conversion of `A => Array[Byte]` can be done using existing `Process` combinators as suggested earlier by @pchlupacek. I don't think in general you will gain much efficiencies with the `logicalLength` and `copy` function approach. You are just pushing the buffering burden onto the callers. 
For example, calculating the `logicalLength` of `A` may not be possible without actually serializing to an array first (or some other dynamic calculation). Also, this has to be called for every emitted value because it cannot be guaranteed it will be the same logical length from one value to the next. 
Implementing `copy` function could potentially require buffering as well since asking for the "last x bytes of the serialized representation of A" would in general require serializing the value first and doing an array copy anyways.
Finally, is there a worry about the mutability of using `Array[Byte]`? The assumption in the implementation is that the array in each emit is treated as immutable, but if there was an `unsafeChunkR` in the pipeline passed into this, wouldn't that cause an issue? Granted, this is an obviously stupid thing to do and out of control of the implementation.